src/runtime/vulkan/vulkan.cc

```diff
@@ -60,35 +91,21 @@ class VulkanThreadEntry {
     pool.reset();
     streams_.clear();
     for (const auto& kv : staging_buffers_) {
-      if (!kv.second) {
-        continue;
-      }
-      auto& buf = *(kv.second);
-      if (buf.host_addr != nullptr) {
-        vkUnmapMemory(buf.device, buf.memory);
-      }
-      if (buf.memory != VK_NULL_HANDLE) {
-        vkFreeMemory(buf.device, buf.memory, nullptr);
-      }
-      if (buf.buffer != VK_NULL_HANDLE) {
-        vkDestroyBuffer(buf.device, buf.buffer, nullptr);
-      }
+      DeleteHostVisibleBuffer(kv.second.get());
```

Comment:
we need to trigger Synchronize when deleting an old Array(in the caller), stream memcpy does that explicitly https://github.com/apache/tvm/blob/e8369881ef69a46eaafd143a1c9bfa0eba2eb003/src/runtime/vulkan/vulkan.cc#L362

This is to ensure that old tasks that uses the smaller staging UBO can get flushed in deferred mode.

Reply:
This change is only refactoring, it doesn't change what the current code does with staging buffer during `VulkanThreadEntry` destructor and `StagingBuffer` reallocation. I just did the refactoring to support the same destruction + reallocation for UBO.

Are you saying we need to add `Synchronize` in the existing code? Also, what about `memcpy` of pack_args to UBO as in https://github.com/apache/tvm/blob/e8369881ef69a46eaafd143a1c9bfa0eba2eb003/src/runtime/vulkan/vulkan.cc#L1158, do we need `Synchronize` there as well?

Reply:
The main point is that we only need to synchronize when re-allocation happens. When we get to a stable state of the UBO staging buffer being big enough, then we won't need to synchronize.

Reply:
BTW, we might want to confirm what is the behavior of writing to the same piece of memory from the host after a previous pipeline that reads the same memory has been submitted, my hope is that this should be fine.

Reply:
ok, UBO reallocation is supposed to happen only pipeline creation time, so it is not frequent I think.

I'm adding code like this, the problem is: What should the cpu device ID be? Can we assume it is 0?
```
  if (buf.device != nullptr && buf.size < size) {
    // free previous buffer
    DeleteHostVisibleBuffer(&buf);
    if (sync_after_realloc) {
      auto cpu_device_id = ???
      VulkanThreadEntry::ThreadLocal()->Stream(cpu_device_id)->Synchronize();
    }
  }
```

Reply:
> BTW, we might want to confirm what is the behavior of writing to the same piece of memory from the host after a previous pipeline that reads the same memory has been submitted

Interesting, this was not an issue when I was making UBO per pipeline. We could still do per pipeline UBO, per thread entry if this is problematic.

Reply:
same problem can appear if we are passing different parameters to the same pipeline(kernel), although being less frequent

Reply:
For the stream, we should record the device stream that is being used for allocating the buffer

Reply:
When allocating a buffer, I only have GPU device ID. But I need a CPU ID for sync. 

Sync on staging buffer has a CPU ID info available to it because it is one of arguments to `CopyDataFromTo`. Note that the existing code that does `Synchronize` on staging buffer does sync on CPU, below `dev_from` is CPU
https://github.com/apache/tvm/blob/e8369881ef69a46eaafd143a1c9bfa0eba2eb003/src/runtime/vulkan/vulkan.cc#L362

Reply:
Interesting, i believe we should sync the GPU id in this case, since stream is being attached to GPU

Reply:
Yes it is likely a bug. The code actually uses the id to look up only GPU context (`VulkanDeviceAPI`, see below), so it doesn't make sense to pass a CPU device ID here. It was working just because both IDs are 0.

https://github.com/apache/tvm/blob/e8369881ef69a46eaafd143a1c9bfa0eba2eb003/src/runtime/vulkan/vulkan.cc#L1116-L1119

I've fixed above points and added sync after UBO realloc in https://github.com/apache/tvm/pull/7833/commits/9df431ff4480fac3edd9422bdb4af2c34a12f5f8